ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+package complete
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+	"golang.org/x/sync/semaphore"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+	realWAL "github.com/onflow/flow-go/ledger/complete/wal"
+	"github.com/onflow/flow-go/module/lifecycle"
+	"github.com/onflow/flow-go/module/observable"
+)
+
+type WALTrieUpdate struct {
+	Update   *ledger.TrieUpdate
+	ResultCh chan<- error
+	TrieCh   <-chan *trie.MTrie
+}
+
+type checkpointResult struct {
+	num int
+	err error
+}
+
+type Compactor struct {
+	checkpointer       *realWAL.Checkpointer
+	wal                realWAL.LedgerWAL
+	checkpointQueue    *CheckpointQueue
+	logger             zerolog.Logger
+	lm                 *lifecycle.LifecycleManager
+	observers          map[observable.Observer]struct{}
+	checkpointDistance uint
+	checkpointsToKeep  uint
+	stopCh             chan chan struct{}
+	trieUpdateCh       <-chan *WALTrieUpdate
+}
+
+func NewCompactor(
+	l *Ledger,
+	w realWAL.LedgerWAL,
+	logger zerolog.Logger,
+	checkpointCapacity uint,
+	checkpointDistance uint,
+	checkpointsToKeep uint,
+) (*Compactor, error) {
+	if checkpointDistance < 1 {
+		checkpointDistance = 1
+	}
+
+	checkpointer, err := w.NewCheckpointer()
+	if err != nil {
+		return nil, err
+	}
+
+	trieUpdateCh := l.TrieUpdateChan()
+	if trieUpdateCh == nil {
+		return nil, errors.New("failed to get valid trie update channel from ledger")
+	}
+
+	tries, err := l.Tries()
+	if err != nil {
+		return nil, err
+	}
+
+	checkpointQueue := NewCheckpointQueueWithValues(checkpointCapacity, tries)
+
+	return &Compactor{
+		checkpointer:       checkpointer,
+		wal:                w,
+		checkpointQueue:    checkpointQueue,
+		logger:             logger,
+		stopCh:             make(chan chan struct{}),
+		trieUpdateCh:       trieUpdateCh,
+		observers:          make(map[observable.Observer]struct{}),
+		lm:                 lifecycle.NewLifecycleManager(),
+		checkpointDistance: checkpointDistance,
+		checkpointsToKeep:  checkpointsToKeep,
+	}, nil
+}
+
+func (c *Compactor) Subscribe(observer observable.Observer) {
+	var void struct{}
+	c.observers[observer] = void
+}
+
+func (c *Compactor) Unsubscribe(observer observable.Observer) {
+	delete(c.observers, observer)
+}
+
+// Ready periodically fires Run function, every `interval`
+func (c *Compactor) Ready() <-chan struct{} {
+	c.lm.OnStart(func() {
+		go c.run()
+	})
+	return c.lm.Started()
+}
+
+func (c *Compactor) Done() <-chan struct{} {
+	c.lm.OnStop(func() {
+		// Signal Compactor goroutine to stop
+		doneCh := make(chan struct{})
+		c.stopCh <- doneCh
+
+		// Wait for Compactor goroutine to stop
+		<-doneCh
+
+		// Shut down WAL component.
+		<-c.wal.Done()
```

Comment:
> should Compactor stop WAL?

Excellent question which requires more than a simple yes/no.

Unless I'm mistaken, node components used to shutdown in sequence but that appears to have been changed to shutdown in parallel.

`WAL` used to be a separate component, at the same level as `Ledger` and `Compactor`.  Node components used to shutdown in reverse sequence so `WAL` shuts down after `Compactor` and `Ledger`.  However, I recently discovered node component shutdown are parallel (not sequential anymore).  Components receive signal and shuts down in parallel, which can cause WAL component to close segment file before all writes are done.

Given this, I removed `WAL` as an independent component, so it needs to be shutdown by either `Ledger` or `Compactor`.  It would be safer for `Compactor` to shutdown `WAL` after all writes are done because `Compactor` drains and processes remaining trie update (writing to WAL segment) in channel after receiving shutdown signal.

> Its supplied as a parameter so I think its lifecycle should be managed by the creator, not here.

Yeah, I was reluctant for the same reason and I used to have `Ledger` shutdown `WAL` in earlier commit.  But that requires communication between `Ledger` and `Compactor` during shutdown because `Compactor` may need to drain and process remaining trie update.  

Thoughts? @m4ksio @ramtinms @zhangchiqing

Reply:
Agree with @fxamacker . 

Make sense to me to stop the WAL here. Better to add some comment to explain
```suggestion
		// only shut down wal after compactor has been shut down, in case there 
		// is still some goroutine writing to WAL files. 
		<-c.wal.Done()
```

Reply:
Ok, we should definitely have an ability to do synchronized shutdowns, but this PR certainly isn't a place to dwell over that, and this solution is fine by me